sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -186,6 +187,15 @@ case class DataSource(
         val path = caseInsensitiveOptions.getOrElse("path", {
           throw new IllegalArgumentException("'path' is not specified")
         })
+        val isSchemaInferenceEnabled = sparkSession.conf.get(SQLConf.STREAMING_SCHEMA_INFERENCE)
+        val isTextSource = providingClass == classOf[text.DefaultSource]
+        // If the schema inference is disabled, only text sources require schema to be specified
+        if (!isSchemaInferenceEnabled && !isTextSource && userSpecifiedSchema.isEmpty) {
+          throw new IllegalArgumentException(
+            "Schema must be specified for creating a streaming source DataFrame. " +
+              "If some input data already exists in the directory, you can create " +
+              "a static DataFrame with spark.read.load(path) and infer schema from it.")
```

Comment:
This should work for ORC, and JSON as well. I just want the user to not get stuck with the thought "ooops, I have to manually create a StructType object!?!?". 

I can make it this string more clear saying, 

"If some files already exists in the directory, then depending on the file format you may be able to create a static DataFrame on that directory with 'spark.read.load(directory)' and infer schema from it."